File: android/core/src/main/kotlin/app/fihaven/core/logic/Reconcile.kt
package app.fihaven.core.logic

import app.fihaven.core.model.SpendTransaction
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.math.abs

/**
 * Bank-vs-manual transaction reconciliation. FiHaven is manual-first; a linked
 * bank (Plaid) adds transactions tagged source="plaid" ALONGSIDE the manual
 * ones, never replacing them — so the same purchase can appear twice. This
 * finds those overlaps to audit, plus bank rows with no manual match. Matching
 * is conservative: same amount (to the cent), a similar merchant, and a date
 * within ±1 day. A suggestion only. Mirrors the web `reconcile.js` (and iOS
 * `Reconcile.swift`).
 */
object Reconcile {
    private fun normMerchant(s: String): String = s.lowercase().filter { it.isLetterOrDigit() }

    /** Do two transactions look like the SAME purchase? */
    fun looksSame(a: SpendTransaction, b: SpendTransaction, dayTolerance: Int = 1): Boolean {
        if (abs(a.amount - b.amount) > 0.01) return false
        val am = normMerchant(a.merchant)
        val bm = normMerchant(b.merchant)
        if (am.length < 3 || bm.length < 3) return false
        if (!am.contains(bm) && !bm.contains(am)) return false
        val da = DateLogic.parseDate(a.date) ?: return false
        val db = DateLogic.parseDate(b.date) ?: return false
        return abs(ChronoUnit.DAYS.between(da, db)).toInt() <= dayTolerance
    }

    data class DuplicatePair(val manual: SpendTransaction, val bank: SpendTransaction)

    /** Pairs where a bank transaction duplicates a manual one (each row paired
     *  at most once, newest bank first) — the audit queue. */
    fun duplicatePairs(transactions: List<SpendTransaction>, dayTolerance: Int = 1): List<DuplicatePair> {
        val manual = transactions.filter { it.source != "plaid" }
        val bank = transactions.filter { it.source == "plaid" }.sortedByDescending { it.date }
        val usedManual = mutableSetOf<String>()
        val pairs = mutableListOf<DuplicatePair>()
        for (b in bank) {
            val m = manual.firstOrNull { !usedManual.contains(it.id) && looksSame(it, b, dayTolerance) }
            if (m != null) { usedManual.add(m.id); pairs.add(DuplicatePair(m, b)) }
        }
        return pairs
    }

    /** Bank transactions with no manual counterpart, newest first. */
    fun unmatchedBank(transactions: List<SpendTransaction>, dayTolerance: Int = 1): List<SpendTransaction> {
        val dupIds = duplicatePairs(transactions, dayTolerance).map { it.bank.id }.toSet()
        return transactions.filter { it.source == "plaid" && it.id !in dupIds }.sortedByDescending { it.date }
    }

    /** Recent manual transactions (within [staleDays], default 35) the bank
     *  never corroborated — ones it "seems to be missing". Newest first. */
    fun unconfirmedManual(transactions: List<SpendTransaction>, today: LocalDate, staleDays: Long = 35): List<SpendTransaction> {
        val cutoff = today.minusDays(staleDays)
        val bank = transactions.filter { it.source == "plaid" }
        return transactions.filter { t ->
            if (t.source == "plaid") return@filter false
            val d = DateLogic.parseDate(t.date) ?: return@filter false
            if (d.isBefore(cutoff) || d.isAfter(today)) return@filter false
            bank.none { looksSame(t, it) }
        }.sortedByDescending { it.date }
    }
}

File: ios/FiHavenCore/Sources/FiHavenCore/Logic/Reconcile.swift
import Foundation

/// Bank-vs-manual transaction reconciliation. FiHaven is manual-first; a linked
/// bank (Plaid) adds transactions tagged source:"plaid" ALONGSIDE the manual
/// ones, never replacing them — so the same purchase can appear twice. This
/// finds those overlaps to audit, plus bank rows with no manual match. Matching
/// is conservative: same amount (to the cent), a similar merchant, and a date
/// within ±1 day. A suggestion only. Mirrors the web `reconcile.js` (and
/// Android `Reconcile.kt`).
public enum Reconcile {
    private static func normMerchant(_ s: String) -> String {
        String(s.lowercased().unicodeScalars.filter { CharacterSet.alphanumerics.contains($0) })
    }

    /// Do two transactions look like the SAME purchase?
    public static func looksSame(_ a: SpendTransaction, _ b: SpendTransaction, tz: TimeZone, dayTolerance: Int = 1) -> Bool {
        if abs(a.amount - b.amount) > 0.01 { return false }
        let am = normMerchant(a.merchant)
        let bm = normMerchant(b.merchant)
        if am.count < 3 || bm.count < 3 { return false }
        if !am.contains(bm) && !bm.contains(am) { return false }
        guard let da = DateLogic.parseDate(a.date, tz: tz),
              let db = DateLogic.parseDate(b.date, tz: tz) else { return false }
        let cal = DateLogic.calendar(tz: tz)
        let days = abs(cal.dateComponents([.day], from: da, to: db).day ?? 99)
        return days <= dayTolerance
    }

    public struct DuplicatePair: Equatable { public let manual: SpendTransaction; public let bank: SpendTransaction }

    /// Pairs where a bank transaction duplicates a manual one (each row paired
    /// at most once, newest bank first) — the audit queue.
    public static func duplicatePairs(_ transactions: [SpendTransaction], tz: TimeZone, dayTolerance: Int = 1) -> [DuplicatePair] {
        let manual = transactions.filter { $0.source != "plaid" }
        let bank = transactions.filter { $0.source == "plaid" }.sorted { $0.date > $1.date }
        var usedManual = Set<String>()
        var pairs: [DuplicatePair] = []
        for b in bank {
            if let m = manual.first(where: { !usedManual.contains($0.id) && looksSame($0, b, tz: tz, dayTolerance: dayTolerance) }) {
                usedManual.insert(m.id)
                pairs.append(DuplicatePair(manual: m, bank: b))
            }
        }
        return pairs
    }

    /// Bank transactions with no manual counterpart, newest first.
    public static func unmatchedBank(_ transactions: [SpendTransaction], tz: TimeZone, dayTolerance: Int = 1) -> [SpendTransaction] {
        let dupIds = Set(duplicatePairs(transactions, tz: tz, dayTolerance: dayTolerance).map { $0.bank.id })
        return transactions.filter { $0.source == "plaid" && !dupIds.contains($0.id) }.sorted { $0.date > $1.date }
    }

    /// Recent manual transactions (within `staleDays`, default 35) that the bank
    /// never corroborated — ones it "seems to be missing". Newest first.
    public static func unconfirmedManual(_ transactions: [SpendTransaction], tz: TimeZone, staleDays: Int = 35, now: Date = Date()) -> [SpendTransaction] {
        let cal = DateLogic.calendar(tz: tz)
        let today = DateLogic.today(tz: tz, now: now)
        guard let cutoff = cal.date(byAdding: .day, value: -staleDays, to: today) else { return [] }
        let bank = transactions.filter { $0.source == "plaid" }
        return transactions.filter { t in
            if t.source == "plaid" { return false }
            guard let d = DateLogic.parseDate(t.date, tz: tz), d >= cutoff, d <= today else { return false }
            return !bank.contains { looksSame(t, $0, tz: tz) }
        }.sorted { $0.date > $1.date }
    }
}
